pkg/provider/consulcatalog/consul_catalog.go

```diff
@@ -294,6 +357,132 @@ func contains(values []string, val string) bool {
 	return false
 }
 
+func rootsWatchHandler(ctx context.Context, dest chan<- []string) func(watch.BlockingParamVal, interface{}) {
+	return func(_ watch.BlockingParamVal, raw interface{}) {
+		if raw == nil {
+			return
+		}
+
+		v, ok := raw.(*api.CARootList)
+		if !ok || v == nil {
+			log.FromContext(ctx).Errorf("invalid result for root certificate watcher")
+			return
+		}
+
+		roots := make([]string, 0, len(v.Roots))
+		for _, root := range v.Roots {
+			roots = append(roots, root.RootCertPEM)
+		}
+
+		dest <- roots
+	}
+}
+
+type keyPair struct {
+	cert string
+	key  string
+}
+
+func leafWatcherHandler(ctx context.Context, dest chan<- keyPair) func(watch.BlockingParamVal, interface{}) {
+	return func(_ watch.BlockingParamVal, raw interface{}) {
+		if raw == nil {
+			return
+		}
+
+		v, ok := raw.(*api.LeafCert)
+		if !ok || v == nil {
+			log.FromContext(ctx).Errorf("invalid result for leaf certificate watcher")
+			return
+		}
+
+		dest <- keyPair{
+			cert: v.CertPEM,
+			key:  v.PrivateKeyPEM,
+		}
+	}
+}
+
+func (p *Provider) createConnectTLSWatchers() (*watch.Plan, *watch.Plan, error) {
+	leafWatcher, err := watch.Parse(map[string]interface{}{
+		"type":    "connect_leaf",
+		"service": p.ServiceName,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create leaf cert watcher plan: %w", err)
+	}
+
+	rootWatcher, err := watch.Parse(map[string]interface{}{
+		"type": "connect_roots",
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create root cert watcher plan: %w", err)
+	}
+
+	return leafWatcher, rootWatcher, nil
+}
+
+func (p *Provider) watchConnectTLS(ctx context.Context, leafWatcher *watch.Plan, rootWatcher *watch.Plan) {
+	ctxLog := log.With(ctx, log.Str(log.ProviderName, "consulcatalog"))
+	logger := log.FromContext(ctxLog)
+
+	leafChan := make(chan keyPair)
+	rootChan := make(chan []string)
+
+	leafWatcher.HybridHandler = leafWatcherHandler(ctx, leafChan)
+	rootWatcher.HybridHandler = rootsWatchHandler(ctx, rootChan)
+
+	logOpts := &hclog.LoggerOptions{
+		Name:       "consulcatalog",
+		Level:      hclog.LevelFromString(logrus.GetLevel().String()),
+		JSONFormat: true,
+	}
+
+	hclogger := hclog.New(logOpts)
+
+	go func() {
+		err := leafWatcher.RunWithClientAndHclog(p.client, hclogger)
+		if err != nil {
+			// This should never be triggered, RunWithClientAndHclog does not seem to ever return anything but nil
```

Comment:
I believe you, but there's no guarantee that will always be the case, and that would force us to constantly re-audit RunWithClientAnHclog to be sure. So I still think we need to take that error into account, just in case.
If I wanted to be lazy about it, I'd panic, but we can't do that in production code.
So yeah, I think we need to make Provide aware of that error as well. WDYT?

Reply:
Makes sense, I'll see what I can do. I do not really like the idea of sending down an error with the certInfo -- but maybe that is the best option. That said it is not yet 100% clear to me how the caller should handle this, given that no error is ever returned currently I am not sure if a abort or a retry would be better…

Reply:
btw, if you're not comfortable passing the error through the certChan, we can also use a (new) dedicated channel for it. Either a new field of Provider, or a new argument to watchConnectTLS. It seems overkill to me, but it's not completely ridiculous.

Reply:
I think it's ok to treat it as a critical error for now (no retry). Let's see how the whole thing looks/works like that, and we can iterate over it afterwards.

Reply:
Ok, done, I hope I did it correctly.

Reply:
Mhm I am now seeing this in semaphore:
```
runtime/debug.Stack(0xc0000e6000, 0x378c7ca, 0x17)\
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f\
github.com/traefik/traefik/v2/pkg/safe.defaultRecoverGoroutine(0x3127f80, 0x55ad140)\
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:66 +0xb4\
github.com/traefik/traefik/v2/pkg/safe.GoWithRecover.func1.1(0x38aa260)\
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:56 +0x57\
panic(0x3127f80, 0x55ad140)\
	/usr/local/go/src/runtime/panic.go:969 +0x1b9\
github.com/traefik/traefik/v2/pkg/provider/consulcatalog.(*Provider).watchConnectTLS(0xc000126e00, 0x3cf2720, 0xc000446c40, 0xc00099a000, 0xc00099a0e0)\
	/go/src/github.com/traefik/traefik/pkg/provider/consulcatalog/consul_catalog.go:475 +0x5b1\
github.com/traefik/traefik/v2/pkg/provider/consulcatalog.(*Provider).Provide.func1(0x3cf2720, 0xc000446c40)\
	/go/src/github.com/traefik/traefik/pkg/provider/consulcatalog/consul_catalog.go:122 +0x50\
github.com/traefik/traefik/v2/pkg/safe.(*Pool).GoCtx.func1()\
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:36 +0x64\
github.com/traefik/traefik/v2/pkg/safe.GoWithRecover.func1(0x38aa260, 0xc0003d2820)\
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:59 +0x4f\
created by github.com/traefik/traefik/v2/pkg/safe.GoWithRecover\
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:53 +0x49\
```
I am not exactly sure what is nil here -- can you help out? I also couldn't reproduce locally.

Reply:
Still failing, new trace with matching line numbers:
```
time=\"2021-02-25T17:54:00Z\" level=error msg=\"Error in Go routine: runtime error: invalid memory address or nil pointer dereference\"
time=\"2021-02-25T17:54:00Z\" level=error msg=\"Stack: goroutine 27 [running]:\
runtime/debug.Stack(0xc0000e6000, 0x378c7ca, 0x17)\
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f\
github.com/traefik/traefik/v2/pkg/safe.defaultRecoverGoroutine(0x3127f80, 0x55ad140)\
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:66 +0xb4\
github.com/traefik/traefik/v2/pkg/safe.GoWithRecover.func1.1(0x38aa260)\
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:56 +0x57\
panic(0x3127f80, 0x55ad140)\
	/usr/local/go/src/runtime/panic.go:969 +0x1b9\
github.com/traefik/traefik/v2/pkg/provider/consulcatalog.(*Provider).watchConnectTLS(0xc00020e280, 0x3cf2720, 0xc000620100, 0xc000560540, 0xc000560620)\
	/go/src/github.com/traefik/traefik/pkg/provider/consulcatalog/consul_catalog.go:475 +0x5b1\
github.com/traefik/traefik/v2/pkg/provider/consulcatalog.(*Provider).Provide.func1(0x3cf2720, 0xc000620100)\
	/go/src/github.com/traefik/traefik/pkg/provider/consulcatalog/consul_catalog.go:122 +0x50\
github.com/traefik/traefik/v2/pkg/safe.(*Pool).GoCtx.func1()\
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:36 +0x64\
github.com/traefik/traefik/v2/pkg/safe.GoWithRecover.func1(0x38aa260, 0xc000448700)\
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:59 +0x4f\
created by github.com/traefik/traefik/v2/pkg/safe.GoWithRecover\
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:53 +0x49\
```

Reply:
Fixed, `connectCert` was nil, but I have no idea why it didn't complain on my machine.